### PR TITLE
Account for CMAC difference in driver-lora

### PR DESCRIPTION
**Pull request by maharjanaman**
Resolves https://github.com/NubeIO/module-core-loraraw/issues/34

### CODE REVIEW

**Comment by maharjanaman** [comment]
@Shiny380 I've tested the changes for MicroEdge and Droplet points but I couldn't test for Ziphydrotap and rubix properly. I tried payload for rubix from [this comment](https://github.com/NubeIO/module-core-loraraw/pull/27#issuecomment-2164316216) but gave me an invalid payload as I assume the payload didn't have cmac. Please do let me know if anything needs to be changed further.

Also, while testing for rubix, only rssi and snr points were added automatically when the device was added.

**Comment by Shiny380** [comment]
Thanks Aman :) I'll get Tan to do testing for Rubix device as he has access to those

File: decoder/rubix.go [hunk 1]
```diff
@@ -546,9 +529,9 @@ func GetRubixPointNames() []string {
 }
 
 func CheckPayloadLengthRubix(data string) bool {
```
Comment (Shiny380) [inline decoder/rubix.go:531]: can you please move this function to the core code somewhere. call it `checkLoRaRAWPayloadLength`. This section is unrelated to the rubix-data-encoding and actually relates to the LoRaRAW protocol which is core to this module.

This function will simply return `true` now

File: decoder/rubix.go [hunk 2]
```diff
@@ -546,9 +529,9 @@ func GetRubixPointNames() []string {
...
-	// 4 bytes address | 1 byte opts | 1 byte nonce | 1 byte length | 1 byte rssi | 1 byte snr
+	// 4 bytes address | 1 byte opts | 1 byte nonce | 1 byte length | 4 byte cmac | 1 byte rssi | 1 byte snr
 	payloadLength := len(data) / 2
-	payloadLength -= 9
+	payloadLength -= 13
 	dataLength, _ := strconv.ParseInt(data[12:14], 16, 0)
```
Comment (Shiny380) [inline decoder/rubix.go:535]: replace this `dataLength` with
```
dataLength := getInnerPayloadLength(data)
```
as seen in below comment

File: pkg/app.go [hunk 3]
```diff
@@ -108,10 +108,27 @@ func (m *Module) handleSerialPayload(data string) {
...
 	}
+	if !devDesc.CheckLength(data) {
+		log.Errorln("invalid payload")
+		return
+	}
+	dataLen := len(data)
+	originalData := data
+	expectedMod := decoder.LoraRawHeaderLen + decoder.LoraRawInnerHeaderLen + decoder.LoraRawCmacLen + decoder.RssiLen + decoder.SnrLen
+	if (dataLen/2)%16 == expectedMod {
+		data = data[14 : dataLen-12]
```
Comment (Shiny380) [inline pkg/app.go:119]: change this to be
```
data = data[14 : getInnerPayloadLength(data)]
```
```
func getInnerPayloadLength(data string) int {
    dataLength, _ := strconv.ParseInt(data[12:14], 16, 0)
    return dataLength
}
```
This removes any padding that existed from the encryption

File: pkg/app.go [hunk 4]
```diff
@@ -108,10 +108,27 @@ func (m *Module) handleSerialPayload(data string) {
...
 		return
 	}
+	if !devDesc.CheckLength(data) {
+		log.Errorln("invalid payload")
+		return
+	}
+	dataLen := len(data)
+	originalData := data
+	expectedMod := decoder.LoraRawHeaderLen + decoder.LoraRawInnerHeaderLen + decoder.LoraRawCmacLen + decoder.RssiLen + decoder.SnrLen
+	if (dataLen/2)%16 == expectedMod {
```
Comment (Shiny380) [inline pkg/app.go:118]:
```
if (dataLen/2)%16 == expectedMod {
    // if(checkLoRaRAWPayloadLength(data) != true
        // return some error
    // data = data[14 : getInnerPayloadLength(data)]
}
if !devDesc.CheckLength(data) {
...
```

**Review by Shiny380 — CHANGES_REQUESTED** [review]
@maharjanaman see requested changes :)

**Comment by Shiny380** [comment]
@lmtan91 will do testing on the rubix decoder

**Review by Shiny380 — APPROVED** [review]

**Pull request merged**